torch/nested/_nested_tensor.py

```diff
@@ -0,0 +1,180 @@
+import operator
+
+# Python tensor subclass version of NT.
+# Properties:
+#  * Constrained to contiguous-only
+#  * Returned for fake / meta conversion
+#  * Supports size() / stride() -> [SymInt] for dynamic shapes
+#  * Supports size() / stride() -> [Union[int, List[int]]] for non-dynamic shapes
+class NestedTensor(Tensor):
+    @staticmethod
+    def __new__(cls, buffer, *args, **kwargs):
+        # Put in obvious garbage; we don't ever want to actually hit the C++ side of this thing
+        return Tensor._make_subclass(cls, torch.empty(0))
+
+    # sizes should be a list of SymInts for the dynamic shapes case
+    def __init__(self, buffer: Tensor, sizes: List[Union[int, List[int], SymInt]]):
+        self.buffer = buffer
+        self.sizes = sizes
+        self.n_tensors = self.sizes[0]
+        self.total_dim = len(self.sizes)
+
+    @classmethod
+    def _sizes_from_shape_list(self, shape_list: Union[Tensor, List[torch.Size]]):
+        # Produces a O(dim) size description from either a Tensor or list of shapes
+        if isinstance(shape_list, Tensor):
+            # Convert Tensor -> list of shapes
+            shape_list = [torch.Size(t.tolist()) for t in shape_list.unbind()]
+
+        if len(shape_list) == 0:
+            return [0]
+
+        rank = len(shape_list[0])
+        for s in shape_list:
+            if len(s) != rank:
+                raise RuntimeError(
+                    "Cannot determine NT size for tensor components of inconsistent rank.")
+
+        sizes = [len(shape_list)]
+        for i in range(rank):
+            dim_sizes = [shape[i] for shape in shape_list]
+            if all([ds == dim_sizes[0] for ds in dim_sizes]):
+                # Consistent value for this dim
+                sizes.append(dim_sizes[0])
+            else:
+                sizes.append(dim_sizes)
+
+        return sizes
+
+    @classmethod
+    def from_tensor_list(
+            cls, tensor_list: List[Tensor], *, dtype=None, device=None, requires_grad=False):
+        sizes = cls._sizes_from_shape_list([t.shape for t in tensor_list])
+        buffer = (
+            # TODO: Support pin_memory
+            torch.cat([t.to(dtype=dtype, device=device).contiguous().view(-1)
+                       for t in tensor_list]).requires_grad_(requires_grad)
+            if len(tensor_list) > 0
+            else torch.tensor([], dtype=dtype, device=device)
+        )
+
+        return cls(buffer, sizes).requires_grad_(requires_grad)
+
+    def __repr__(self):
+        prefix = "NestedTensor"
+        if self.is_meta or isinstance(self.buffer, torch._subclasses.FakeTensor):
+            return f"{prefix}(..., n_tensors={self.n_tensors}, dim={self.total_dim}, device=\"{self.device}\")"
+
+        s = self.unbind()
+        list_str = '\n'.join([str(t) for t in s])
+        return f"{prefix}([\n{list_str}\n])"
+
+    def dim(self):
+        return self.total_dim
+
+    # nt.shape calls here as well
+    def size(self, dim=None):
```

Comment:
NB: probably need `stride()` too for downstream, and we can easily auto-generate it from `sizes` because of contiguous constraint